### PR TITLE
Update some of the Python dependencies from Beam Dependency Check Report

### PR DESCRIPTION
**Pull request by kamilwu**
Dependencies updated:
- fastavro
- cachetools
- oauth2client

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kamilwu** [comment]
Run Python 3.8 PostCommit

**Comment by kamilwu** [comment]
R: @chamikaramj

**Comment by kamilwu** [comment]
R: @tvalentyn

File: sdks/python/apache_beam/io/avroio.py [hunk 1]
```diff
@@ -627,11 +627,19 @@ def write_record(self, writer, value):
     writer.append(value)
 
 
+class _FastAvroWriter(Writer):
+  """An adapter class which exposes a file handle so that it can be closed
+  by the sink. """
+  def __init__(self, file_handle, schema, codec):
+    super(_FastAvroWriter, self).__init__(file_handle, schema, codec)
```
Comment (tvalentyn) [inline sdks/python/apache_beam/io/avroio.py:634]: What is the context for this change (do you by chance have a pertaining fastavro commit that changes this behavior)?

File: sdks/python/setup.py [hunk 2]
```diff
@@ -141,7 +141,7 @@ def get_version():
     # server, therefore list of allowed versions is very narrow.
     # See: https://github.com/uqfoundation/dill/issues/341.
     'dill>=0.3.1.1,<0.3.2',
-    'fastavro>=0.21.4,<0.24',
+    'fastavro>=1.0.0,<2',
```
Comment (tvalentyn) [inline sdks/python/setup.py:144]: Do we need to increase the lower bound (is the change not compatible with earlier fastavro releases)?

Comment (kamilwu) [reply sdks/python/setup.py]: Not really. I'll restore the previous bound

**Comment by kamilwu** [comment]
Run Python 3.8 PostCommit

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (kamilwu) [reply sdks/python/apache_beam/io/avroio.py]: It turned out this change is unnecessary -- I'll remove it then.

The problem was with pyenv and Python interpreter on my machine. It lacked libpython.so. As a result, I couldn't import cythonized code. Instead, a pure Python version, `_write_py.py`, was being used: https://github.com/fastavro/fastavro/blob/master/fastavro/write.py#L3. That pure Python version has some differences, but most importantly, the `Writer` class doesn't set `self.fo` field: 

https://github.com/fastavro/fastavro/blob/master/fastavro/_write_py.py#L411 (pure Python)
vs.
https://github.com/fastavro/fastavro/blob/master/fastavro/_write.pyx#L559 (cython)

I wrongly assumed it was the fault of updating the package.

Comment (tvalentyn) [reply sdks/python/apache_beam/io/avroio.py]: Actually, it is a change in 0.22.0: https://github.com/fastavro/fastavro/pull/343/commits/6c55c318735278d097b05da7077b23fead50a039#diff-96a4a8649900b3ff9f22a57de25baca4R498
https://github.com/fastavro/fastavro/blob/f468264b0f86661f1893c8b32bc592fb44e39c75/ChangeLog#L91

Cython is an optional dependency in Beam, so we should not have to rely on presence of cython.

It seems strange that fastavro public api is now different in cythonized and non-cythonized version. Are you interested in reaching out to fastavro folks to clarify whether this is something they want to fix?

We explicitly close the filehandle due to: https://issues.apache.org/jira/browse/BEAM-7747. Perhaps there is a better way?

Comment (tvalentyn) [reply sdks/python/apache_beam/io/avroio.py]: Asked on https://github.com/fastavro/fastavro/pull/343#discussion_r497095926

Comment (kamilwu) [reply sdks/python/apache_beam/io/avroio.py]: Thanks. We've already got an answer, let's wait until we know if it's possible for the `fo` attribute to become available.

Comment (tvalentyn) [reply sdks/python/apache_beam/io/avroio.py]: Let's do the following:

1. File a JIRA for this issue, mention in the the error message that you observed. Let's mention in the JIRA  the workaround (install cython or downgrade fastavro to 0.21.24) in case users discover this error.

2. Add a workaround in code:
```
try:
  writer.fo.close()
except AttributeError:
  # We can clean this up after fastavro lower bound includes  a fix for https://github.com/fastavro/fastavro/issues/472.
  writer.encoder.close()
```
3. Test that the workaround works.

Thanks for spotting this.  How did you catch it by the way? Was it a unit test failure that occurred locally? I wonder why our non-cython test suite does not catch it.

Comment (kamilwu) [reply sdks/python/apache_beam/io/avroio.py]: > Was it a unit test failure that occurred locally?

Yes.

> I wonder why our non-cython test suite does not catch it.

I think it doesn't really depend on whether cython is installed or not. `pip install fastavro` downloads a wheel built on the specific platform, which already contains libraries that can be imported. For example, on my Linux machine:

```
pip install --force-reinstall fastavro==0.23.6

Collecting fastavro==0.23.6
  Using cached fastavro-0.23.6-cp37-cp37m-manylinux2010_x86_64.whl (1.4 MB)
```

```
ls site-packages/fastavro | grep write 

json_write.py
_logical_writers.cpython-37m-x86_64-linux-gnu.so
logical_writers.py
_logical_writers_py.py
_write.cpython-37m-x86_64-linux-gnu.so   <-- this is the library I'm talking about
write.py
_write_py.py
```

I'm pretty sure the same happens on Jenkins and on our users' machines. So how did I catch the bug? That `.so` library links with `libpython3.x.so`:

```
ldd site-packages/fastavro/_write.cpython-37m-x86_64-linux-gnu.so

linux-vdso.so.1 (0x00007ffd7abf7000)
libpython3.7m.so.1.0 => not found
libc.so.6 => /lib64/libc.so.6 (0x00007f8c548e3000)
/lib64/ld-linux-x86-64.so.2 (0x00007f8c54d07000)
```
which I didn't have on my machine due to misconfiguration (I already know it's pyenv fault, but this is a different story). That's why the first `import` statement failed, `ImportError` was raised, and pure Python version was used instead.

To sum up, it is a problem, but the chances that someone will encounter this are pretty low. Anyway, I think it still makes sense to file a JIRA and test the workaround that you proposed.

Comment (tvalentyn) [reply sdks/python/apache_beam/io/avroio.py]: Ah that's right. Cython presence only matters if  we are  installing fastavro from sources.
